### PR TITLE
DNM: Thoughts on this since the beginning

### PR DESCRIPTION
**Pull request by ChristinaMeno**
@nthomas-redhat
@balamurugana

I made this PR to be able to share feedback.

### CODE REVIEW

File: usm_rest_api/v1/views/views.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        except Exception, e:
+            log.exception(e)
+            #Delete the node from list to avoid further processing
+            nodelist.remove(node)
+            failedNodes.append(node)
+    #Sleep for sometime to make sure all the restarted minions are back
+    #online
+    log.debug("Accepting the minions keys" )
+    createGlusterCluster.update_state(state='ADD_MINION_KEYS_AND_DB_UPDATE')
+    time.sleep( 3 )
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:203]: This is going to be unstable depending on the load.

File: usm_rest_api/v1/views/views.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+                nodelist.remove(node)
+            continue
+        
+    #Now create the cluster
+    log.debug("peer probe start" )
+    createGlusterCluster.update_state(state='PEER_PROBE')
+    #
+    #Wait for some time so that the communication chanel is ready
+    #
+    time.sleep(10)
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:247]: What are we really waiting for? Could we check that instead?

File: usm_rest_api/v1/views/views.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        #TODO Get the host uuid from  host and update in DB
+        #
+        node['node_id'] = uuid.uuid4()
+        node['cluster'] = str(postdata['cluster_id'])
+        try:
+            saltapi.accept_minion(socket.gethostbyaddr(node['management_ip'])[0])
+        except Exception, e:
+            log.exception(e)
+            #Delete the node from list to avoid peer probing
+            nodelist.remove(node)
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:216]: Why change the list that you are iterating over?

File: usm_rest_api/v1/views/views.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    createGlusterCluster.update_state(state='PEER_PROBE')
+    #
+    #Wait for some time so that the communication chanel is ready
+    #
+    time.sleep(10)
+    #Need peerprobe only if the length is more than one
+    if len(nodelist)>1:
+        rootNode = nodelist[0]['management_ip']
+        del nodelist[0]
+        for node in nodelist:
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:252]: all this modification to nodelist is confusing

File: usm_rest_api/v1/views/views.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    """
+    log.debug("Inside create_cluster. Request Data: %s" % request.data )
+    data = request.data.copy()
+    #
+    #TODO:Check the type of the cluster to be created and call appropriate task
+    #
+    #create gluster cluster
+    jobId = createGlusterCluster.delay(data)
+    log.debug("Exiting create_cluster JobID: %s" % jobId)
+    return Response(str(jobId), status=200)
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:156]: 200 is the wrong status code, you want 202

File: usm_rest_api/v1/views/views.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        #Create a host and add to gluster cluster
+        jobId = createGlusterHost.delay(data)
+        log.debug("Exiting ... JobID: %s" % jobId)
+        return Response(str(jobId), status=200)
+   
+@shared_task
+def createGlusterHost(data):
+    log.debug("Inside createGlusterHost Async Task")
+    #Setup the host communication
+    try:
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/views/views.py:330]: This try: except: clause is far too large.

File: usm_rest_api/v1/urls.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+router = routers.DefaultRouter()
+router.register(r'users', v1_views.UserViewSet)
+router.register(r'clusters', v1_views.ClusterViewSet)
+router.register(r'hosts', v1_views.HostViewSet)
+
+
+urlpatterns = patterns('',
+                       url(r'^auth/login', v1_views.login),
+                       url(r'^auth/logout', v1_views.logout),
+                       url(r'^utils/get_ssh_fingerprint/(.*)/$', v1_views.get_ssh_fingerprint),
```
Comment (ChristinaMeno) [inline usm_rest_api/v1/urls.py:14]: This pattern needs to be more restrictive

File: usm_rest_api/v1/views/views.py (hunk 1 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:203]: This is a temporary arrangement to wait till minion is Up and running. Now Bala has re-factored the Back-end code such that this wait is not required. I will update this once h pushes his code into the master

File: usm_rest_api/v1/views/views.py (hunk 2 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:247]: Salt takes some time to exchange the keys and some other internal stuff. So if I execute the command immediately after accepting the keys, the command fails and hence the delay. So is there a way to know whether the channel is up? One way is to keep pinging the nodes to figure out but don't think that's good either. Can you suggest something?

File: usm_rest_api/v1/views/views.py (hunk 3 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:216]: Even  this implementation also going to change as the failed nodes also added to the DB with a status down.

File: usm_rest_api/v1/views/views.py (hunk 4 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:252]: I am taking out first node from the list as root node and form the cluster with rest of the nodes in the cluster

File: usm_rest_api/v1/views/views.py (hunk 5 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:156]: Yes I will change this

File: usm_rest_api/v1/views/views.py (hunk 6 above)
Comment (cybernth) [inline usm_rest_api/v1/views/views.py:330]: Will modify the code

File: usm_rest_api/v1/urls.py (hunk 7 above)
Comment (cybernth) [inline usm_rest_api/v1/urls.py:14]: Will update the code

**Pull request closed without merging**